cmd/fleet/cron.go

```diff
@@ -745,11 +751,68 @@ func newCleanupsAndAggregationSchedule(
 				return ds.UpdateOSVersions(ctx)
 			},
 		),
+		schedule.WithJob(
+			"verify_disk_encryption_keys",
+			func(ctx context.Context) error {
+				return verifyDiskEncryptionKeys(ctx, logger, ds, config)
+			},
+		),
 	)
 
 	return s, nil
 }
 
+func verifyDiskEncryptionKeys(
+	ctx context.Context,
+	logger kitlog.Logger,
+	ds fleet.Datastore,
+	config *config.FleetConfig,
+) error {
+	if !config.MDM.IsAppleSCEPSet() {
+		logger.Log("inf", "skipping verification of encryption keys as MDM is not fully configured")
+		return nil
+	}
+
+	keys, err := ds.GetUnverifiedDiskEncryptionKeys(ctx)
+	if err != nil {
+		logger.Log("err", "unable to get unverified disk encryption keys from the database", "details", err)
+		return err
+	}
+
+	cert, _, _, err := config.MDM.AppleSCEP()
+	if err != nil {
+		logger.Log("err", "unable to get SCEP keypair to decrypt keys", "details", err)
+		return err
+	}
+
+	parsed, err := x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		logger.Log("err", "unable to get SCEP keypair to decrypt keys", "details", err)
+		return err
+	}
+
+	decryptable := []uint{}
+	undecryptable := []uint{}
+	for _, key := range keys {
+		if _, err := apple_mdm.DecryptBase64CMS(key.Base64Encrypted, parsed, cert.PrivateKey); err != nil {
+			undecryptable = append(undecryptable, key.HostID)
+			continue
+		}
+		decryptable = append(decryptable, key.HostID)
+	}
+
+	if err := ds.SetHostDiskEncryptionKeyStatus(ctx, decryptable, true); err != nil {
```

Comment:
Is it possible that one of the keys in this loop could be updated again while this job is pending? In that case I think the new key could mistakenly be marked as verified based on the old key.

Reply:
That's a great point, it definitely could. We could do the update based on the base64key being the same value as when the test was done, but it would prevent doing the nice (and likely efficient) batch update that `SetHostDiskEncryptionKeyStatus` currently does.

Thinking out loud, if `GetUnverifiedDiskEncryptionKeys` returned the current mysql timestamp as part of its result (or even the `updated_at` of each row, and we can then figure out the latest of those), then I think the `SetHostDiskEncryptionKeyStatus` update could look like this:

```
"UPDATE host_disk_encryption_keys SET decryptable = ? WHERE host_id IN (?) AND updated_at <= ?"
```

passing the timestamp at which the read occurred for the last placeholder `?`. This means that if any key got updated during processing, it would have to be greater than the timestamp. It's not perfect, as mysql's timestamps have terrible resolution (second, I think), but it'd be reasonably safe at a minimal complexity.

Reply:
great catch!! and thanks for the suggestion! I will adjust tomorrow first time my morning.

Reply:
@gillespi314 thanks again for the great catch! Fixed with @mna's suggestion in [5b787fe](https://github.com/fleetdm/fleet/pull/9712/commits/5b787fe04ca4cb079bb3489836aa0a96cd3be3d0)